feat: GitHub Actions docker files CI

nschonni opened the pull request:
Took @LaurentGoderre approach to the update.sh script generating out the Travis-CI file and converted it to create separate GitHub Actions.
- The benefit of this over the Travis diff checking, is the `path` approach limits the matrix to only queue jobs when a particular Dockerfile is touched.
- ~~I added the DO NOT MERGE to this because I included a junk commit at the end to show the matrix getting triggered when the files are touched.~~
- This does not convert the existing deploy setup from Travis-CI
- Needed to remove the `-it` flag from the BATS testing since that doesn't seem to work on Actions
- Does not remove the CI job when a variant/version is removed currently. EX: like the 8 EOL

nschonni commented:
Right now, the targets are only setup to run on `pull_request` and not `push`, so there aren't any tests for the master push.
The only think I think might make sense to go with this would be to enable the repo setting of "branch is up to date". EX: you get the annoying "Update" button when ever a different PR is merged, but then only things tested agains master can land.

LaurentGoderre commented:
With this, we loose the ability to build branches and there's no longer a build on merges?

LaurentGoderre commented:
I like the approahc but I think it still should merge on branches and merges.

nschonni commented:
They rules could be changed to run on master, but I think making the branch "strict" https://help.github.com/en/github/administering-a-repository/types-of-required-status-checks would give the same result. It's less of an issue with hitting the "Update" button with this, since only affected images would rebuilt, rather than all the jobs

nschonni commented:
I've popped off the testing commit so this can land if there is agreement

PeterDaveHello commented on test-image.bats:
```diff
@@ -1,17 +1,17 @@
 
 @test "Test for node and version" {
-  run docker run --rm -it node:"$full_tag" node -e "process.stdout.write(process.versions.node)"
+  run docker run node:"$full_tag" node -e "process.stdout.write(process.versions.node)"
```
I'm not sure how is this related with switching to GitHub Actions?

nschonni replied:
I'm going to hold off changing anything for BATS because it sounds like https://github.com/nodejs/docker-node/pull/1197 might mean it gets dropped from the build

PeterDaveHello replied:
No matter they are necessary or not, they're not related with switching to GitHub Actions, are they?

SimenB replied:
From the OP

>Needed to remove the `-it` flag from the BATS testing since that doesn't seem to work on Actions

`--rm` should be kept regardless, annoying if running locally

SimenB replied:
could keep the `-t` maybe? if it's only the interactive part which doesn't work

nschonni replied:
If BATS gets removed in the other PR as redundant, i'll be removing the BATS install and the changes here

SimenB replied:
I don't wanna remove bats unless somebody feels strongly about it. I think it provides a value by demonstrating node is runnable from the outside of the image build process.

PeterDaveHello replied:
It's okay to keep it but the changes here should be separated into another PR.

nschonni replied:
Added back the `--rm`

PeterDaveHello commented on .github/workflows/10-alpine3.11.yml:
```diff
@@ -0,0 +1,17 @@
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v1
+      - run: sudo apt-get install bats
```
Maybe need to add `-y` here for apt?

nschonni replied:
Doesn't seem to be needed when I was testing

nschonni commented:
@LaurentGoderre I added back the build on `push` for now, since the publish script is relying on the merge commit right now. Also added the build and test scripts as triggers for the testing. It caught my Chakracore mistake, and I've added the patch here if it doesn't land in 13.11 PR first

LongLiveCHIEF commented:
What mechanism are you guys using to actually push images to docker hub?  I've seen you mention travis, but when I look at the script in the travis action for the deploy stage, all it does is build as far as I can tell.

I recently took over as the primary maintainer of the octoprint/octoprint official image, and I'm really enjoying the way you guys are doing things, but I'm just looking to connect the dots for that final step.

Are you doing the final publishing of images via auto builds on docker cloud?

Love your work, and appreciate any insight you can give me about the image publish step of your overall workflow!

SimenB commented:
>What mechanism are you guys using to actually push images to docker hub?

The bot opens a PR to the official images, which ends up pushing to docker hub. E.g. the latest: https://github.com/docker-library/official-images/pull/7697

LongLiveCHIEF commented:
Of course! Can't believe I missed that since you guys are an official image it would have to be in docker-library/official-images repo. Thanks for filling in the blank for me!

LongLiveCHIEF commented on test-image.bats:
```diff
@@ -1,17 +1,17 @@
 #!/usr/bin/env bats
 
 @test "Test for node and version" {
-  run docker run --rm -it node:"$full_tag" node -e "process.stdout.write(process.versions.node)"
```
@nschonni would it work if you only removed the psuedo-tty `-t` but left the `-i`?  If it's in CI, it shouldn't need the tty anyways right?  

`docker run --rm -i node:"$full_tag"`

nschonni replied:
Not sure, but these tests also just overlap what is done in the image build now, so i'm not really concerned

LongLiveCHIEF replied:
yeah, was just curious since i'll soon have to test the same thing, and you already have everything set up :rofl: :thumbsup:

nschonni commented:
@nodejs/docker do you want to land this before v14?

SimenB commented:
I see no reason to wait to land this if we think it works - it shouldn't be observable to consumers anyways.

The way it's set up now it'll trigger an upstream PR even if a build fails - is the idea that it'll fail from the PR first anyways, so no broken versions can land on master?

nschonni commented:
Yes, but i'll also look at that after this lands. I just didn't want to mix too much together

SimenB commented on .github/workflows/10-alpine3.10.yml:
```diff
@@ -0,0 +1,23 @@
+      - 10/alpine3.10/Dockerfile
+
+jobs:
+  build:
```
Seems this name is used in the status? Can we set a display name here?

nschonni replied:
The job is the middle part of the display `10 on alpine3.10 / build (pull_request)`. Th start of the display comes from the `name` above

SimenB replied:
Not on mobile

https://drive.google.com/file/d/1eF3hdszKvhtxdGuoKg8qJK41fK64xoCK/view?usp=drivesdk

(No idea how to upload screenshot from mobile, sorry about the google drive link)

SimenB replied:
On a computer now, so can paste it

![image](https://user-images.githubusercontent.com/1404810/79908321-51d33100-841b-11ea-88c3-14f1c939e62c.png)

SimenB replied:
This is still an issue. @nschonni if you give the job a `name` it works on mobile as well

nschonni replied:
I can't see it on mine, but i've duplicated the root name again

SimenB replied:
works now 👍

SimenB commented:
Needs to be updated for v14.

Should we keep travis for this to try it out, and then remove travis if we're happy? I fully expect we'll be happy with it, but who knows 🙂

nschonni commented:
Added v14, and dropped v13

SimenB commented:
@nodejs/docker thoughts on this one? I like it, FWIW 🙂

SimenB commented on update.sh:
```diff
@@ -17,26 +17,26 @@ function usage() {
   OPTIONS:
     -s Security update; skip updating the yarn and alpine versions.
-    -t Travis CI config update only
+    -t CI config update only
```
`t` or `b`?

SimenB replied:
Seems like it's still `t` based on line 38, just the change on line 20 that inconsistent then?

nschonni replied:
Yeah, I tried changing it to `c` at one point and it just stopped working properly, so I just undid the letter change

nschonni replied:
OK, I figured out my issue with not updating the `getopts`, so it's "b" now

nschonni commented:
I think I'll go ahead an land this sometime after the security releases are out and are successfully published next week

nschonni commented:
I'll land this tomorrow unless anyone from @nodejs/docker has any objections

SimenB commented:
I'm still hugely in favour 👍
